[FSDP] Allow nested FSDP wrapper to use different mixed precision

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90523

The main change is to move `args` and `kwargs` dtype convertion
from `_root_pre_forward` to `_pre_forward`, so that every
FSDP has a chance to apply its own precision.

Differential Revision: [D41883717](https://our.internmc.facebook.com/intern/diff/D41883717)

mrshenli committed:
[FSDP] Allow nested FSDP wrapper to use different mixed precision

[ghstack-poisoned]

mrshenli committed:
Update on "[FSDP] Allow nested FSDP wrapper to use different mixed precision"


The main change is to move `args` and `kwargs` dtype convertion
from `_root_pre_forward` to `_pre_forward`, so that every
FSDP has a chance to apply its own precision.

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -257,12 +266,14 @@ def _pre_forward(
     module: nn.Module,
-    input: Any,
-):
+    args: Tuple[Any],
```
nit: Does this need to be `Tuple[Any, ...]`? (also affects docstring)

My understanding is that `Tuple[Any]` means a singleton tuple with an element of any type and that `Tuple[Any, ...]` means a tuple with arbitrarily many elements, where each element may be any type.

mrshenli replied:
Oh Yep, thanks for the catch!

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -284,6 +296,12 @@ def _pre_forward(
     _register_post_backward_hooks(state, handles)
 
+    # Recursively convert args and kwargs to specified precision.
+    input_dtype: Optional[torch.dtype] = state.mixed_precision.param_dtype
```
I see that this line only targets the wrapper FSDP code path.

I recommend adding a `TODO` at this line to indicate that we need to change it to support the non-wrapper code path; however, we may also file an issue if you do not want to re-run CI.

---
The open problem is how should `fully_shard` enable applying different `fsdp_kwargs` to different `FlatParamHandle` (for a select subset of the arguments). The main difference is that `fully_shard` does not support manual wrapping, which is the solution for `FullyShardedDataParallel`. Today, one application of `fully_shard` does not interact with any other application of `fully_shard`. If we want to migrate to the composable APIs, then we need to iron out the design here.

Given a away to specify different `fsdp_kwargs` to different `FlatParamHandle`, then the change to this line here is not complicated.

For the non-wrapper code path, we change the requirement to be that every handle in `handles` (passed to this `_pre_forward`) has the same low precision parameter dtype. In today's design (and probably for the near future), `handles` is a singleton (i.e. `len(handles) == 1`). Either way, the general constraint will be on `handles`.

```
def _get_handles_param_dtype(handles: Iterable[FlatParamHandle]) -> Optional[torch.dtype]:
    """
    Assumes all ``handles`` have the same low precision parameter dtype
    setting, where ``None`` represents full precision. This returns the single
    low precision parameter dtype setting across ``handles``.
    """
    dtypes: Optional[torch.dtype] = set(handle._config.low_prec_param_dtype for handle in handles)
    p_assert(
        len(dtypes) == 1,
        f"Expects uniform low precision parameter dtype but got {dtypes}",
    )
    return next(iter(dtypes))        
```
```
input_dtype: Optional[torch.dtype] = _get_handles_param_dtype(handles)
```

(The question is how to specify `handle._config._low_prec_param_dtype` to be different across `FlatParamHandle`s of the same `fully_shard`.)

mrshenli replied:
> Today, one application of fully_shard does not interact with any other application of fully_shard. If we want to migrate to the composable APIs, then we need to iron out the design here.

Can we do the following:

For short term, we require users to wrap children modules with `fully_shard` before wrapping parent ones. When applying `fully_shard`, recursively check all submodules' states and skip the ones that already have `replicate`, `fully_shard` on it. More specifically, `fully_shard` can write a flag in every submodule, indicating that this submodule is handled by a parent module already. If users later applied another `fully_shard` on the submodule, error out. 

For the longer term, when we have the high-level wrapper API, that API can sort things out, and we no longer require users to wrap things in order.

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -284,6 +296,12 @@ def _pre_forward(
 
+    # Recursively convert args and kwargs to specified precision.
+    input_dtype: Optional[torch.dtype] = state.mixed_precision.param_dtype
+    args, kwargs = _prepare_forward_inputs(
```
The main assumption on the performance side from this PR is that the overhead from `_prepare_forward_inputs()` / `_cast_fp_inputs_to_dtype()` is not a concern.

Has anyone confirmed if `t.to(t.dtype)` is fully a no-op with effectively 0 overhead? If not, I wonder if we should change `_cast_fp_inputs_to_dtype()` to guard the `.to()` with an `if` statement:
```
def cast_fn(x: torch.Tensor) -> torch.Tensor:
    def cast_fn(x: torch.Tensor) -> torch.Tensor:
    if not torch.is_floating_point(x) or x.dtype == dtype:
        return x
    y = x.to(dtype)
    # Explicitly copy over `requires_grad` since this runs inside
    # `torch.no_grad()`
    if x.is_leaf:
        y.requires_grad = x.requires_grad
    return y
```
In other words, we do not perform any of the logic if the tensor's dtype already matches. I think with this change, I would be more confident that there will be effectively no performance regression from this. The only additional overhead is a CPU-side recursion over the input data structure via `_apply_to_tensors()` incurred for every FSDP instance.

---

On a related note, I think we need to be careful about using a side stream to do this downcasting. Today, we do not use such a stream, but we have mentioned considering to do so. There may be some strange memory freeing behavior and stream complications. We are investigating this for the post-backward, so I think we should hold off on any stream changes in the near term until we get a better understanding.

mrshenli replied:
The implementation of `Tensor.to()` is: 

https://github.com/pytorch/pytorch/blob/eeb3f8aa5496f713339fea678d145575c884e459/aten/src/ATen/core/Tensor.cpp#L41-L51

Then I traced where it calls to.

```c++
struct TORCH_API to_dtype {
  using schema = at::Tensor (const at::Tensor &, at::ScalarType, bool, bool, c10::optional<at::MemoryFormat>);
  using ptr_schema = schema*;
  // See Note [static constexpr char* members for windows NVCC]
  STATIC_CONSTEXPR_STR_INL_EXCEPT_WIN_CUDA(name, "aten::to")
  STATIC_CONSTEXPR_STR_INL_EXCEPT_WIN_CUDA(overload_name, "dtype")
  STATIC_CONSTEXPR_STR_INL_EXCEPT_WIN_CUDA(schema_str, "to.dtype(Tensor(a) self, ScalarType dtype, bool non_blocking=False, bool copy=False, MemoryFormat? memory_format=None) -> Tensor(a)")
  static at::Tensor call(const at::Tensor & self, at::ScalarType dtype, bool non_blocking, bool copy, c10::optional<at::MemoryFormat> memory_format);
  static at::Tensor redispatch(c10::DispatchKeySet dispatchKeySet, const at::Tensor & self, at::ScalarType dtype, bool non_blocking, bool copy, c10::optional<at::MemoryFormat> memory_format);
};
```

```c++
// aten::to.dtype(Tensor(a) self, ScalarType dtype, bool non_blocking=False, bool copy=False, MemoryFormat? memory_format=None) -> Tensor(a)
at::Tensor to_dtype::redispatch(c10::DispatchKeySet dispatchKeySet, const at::Tensor & self, at::ScalarType dtype, bool non_blocking, bool copy, c10::optional<at::MemoryFormat> memory_format) {

    static auto op = create_to_dtype_typed_handle();
    return op.redispatch(dispatchKeySet, self, dtype, non_blocking, copy, memory_format);
}

```

```c++
// aten::to.dtype(Tensor(a) self, ScalarType dtype, bool non_blocking=False, bool copy=False, MemoryFormat? memory_format=None) -> Tensor(a)
static C10_NOINLINE c10::TypedOperatorHandle<to_dtype::schema> create_to_dtype_typed_handle() {
  return c10::Dispatcher::singleton()
      .findSchemaOrThrow(to_dtype::name, to_dtype::overload_name)
      .typed<to_dtype::schema>();
}
```

https://github.com/pytorch/pytorch/blob/eeb3f8aa5496f713339fea678d145575c884e459/aten/src/ATen/core/dispatch/Dispatcher.h#L489-L492

https://github.com/pytorch/pytorch/blob/eeb3f8aa5496f713339fea678d145575c884e459/aten/src/ATen/core/dispatch/Dispatcher.h#L641-L656


At this point, I didn't continue. Even if it does not create any CUDA kernel, the above is sufficient complexity + python/cpp context switch. Let's avoid that by using your suggestion above.

rohan-varma replied:
Agreed, looks like even if tensor.to() is a noop, we can avoid a python/C++ switch and GIL release / reacquire here.

awgu commented on torch/testing/_internal/common_distributed.py:
```diff
@@ -953,3 +954,26 @@ def perThreadTearDown(self):
     def world_size(self) -> int:
         raise RuntimeError("world size not implemented")
+
+
```
I see: So this `CheckPrecisionModule` does not check precision itself -- it is used to check the heterogeneous mixed precision case externally.

mrshenli replied:
Let me update the name to match that.

awgu reviewed (approved):
This PR looks good to me. Feel free to give a chance for @rohan-varma  to review if he wants and also to let me know what you think about some of the comments I left.

mrshenli committed:
Update on "[FSDP] Allow nested FSDP wrapper to use different mixed precision"


The main change is to move `args` and `kwargs` dtype convertion
from `_root_pre_forward` to `_pre_forward`, so that every
FSDP has a chance to apply its own precision.

[ghstack-poisoned]

rohan-varma reviewed (approved):
LGMT!

mrshenli commented on test/distributed/_composable/test_fully_shard.py:
```diff
@@ -407,5 +411,42 @@ def _check_same_param_handles(
                 self.assertTrue(composable_fqn.endswith(wrapped_fqn))
 
 
+class TestMixedPrecision(FSDPTest):
```
Thanks for the catch! Removed the one below.

mrshenli commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -284,6 +296,12 @@ def _pre_forward(
 
+    # Recursively convert args and kwargs to specified precision.
+    input_dtype: Optional[torch.dtype] = state.mixed_precision.param_dtype
+    args, kwargs = _prepare_forward_inputs(
```
Yep, @awgu also pointed this out. Added a condition to skip when the dtype already match.

awgu replied:
Yes, that is what we proposed in this PR, and I believe @mrshenli has added that change to `_cast_fp_inputs_to_dtype()`.

mrshenli commented:
@pytorchbot merge -g

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41883717).
